pulser-core/pulser/backend/default_observables.py

```diff
@@ -0,0 +1,369 @@
+        *,
+        evaluation_times: Sequence[float] | None = None,
+        one_state: Eigenstate | None = None,
+    ):
+        """Initializes the observable."""
+        super().__init__(evaluation_times=evaluation_times)
+        # TODO: Checks
+        self.one_state = one_state
+
+    def name(self) -> str:
+        """Returns the name of the observable."""
+        return "correlation_matrix"
+
+    @staticmethod
+    @functools.cache
+    def _get_number_operator(
+        qudit_ids: frozenset[int],
+        n_qudits: int,
+        eigenstates: Sequence[Eigenstate],
+        one_state: Eigenstate,
+        op_type: Type[OperatorType],
+    ) -> OperatorType:
+        n_op = {one_state * 2: 1.0}
+        return op_type.from_operator_repr(
+            eigenstates=eigenstates,
+            n_qudits=n_qudits,
+            operations=[(1.0, [(n_op, qudit_ids)])],
+        )
+
+    def apply(
+        self, *, state: State, hamiltonian: Operator, **kwargs: Any
+    ) -> list[list]:
+        """Calculates the observable to store in the Results."""
+        try:
+            return state.get_correlation_matrix(one_state=self.one_state)
+        except NotImplementedError:
+            pass
+
+        @functools.cache
+        def calc_expectation(qudit_ids: frozenset[int]) -> Any:
+            return self._get_number_operator(
+                qudit_ids,
+                state.n_qudits,
+                state.eigenstates,
+                self.one_state or state.infer_one_state(),
+                type(hamiltonian),
+            ).expect(state)
+
+        return [
+            [
+                calc_expectation(frozenset((i, j)))
+                for j in range(state.n_qudits)
+            ]
+            for i in range(state.n_qudits)
+        ]
+
+
+class QubitDensity(Observable):  # TODO: Rename
+    """Stores the population of an eigenstate on each qudit.
+
+    Args:
+        evaluation_times: The relative times at which to compute the
+            correlation matrix. If left as `None`, uses the
+            `default_evaluation_times` of the backend's `EmulationConfig`.
+        one_state: The eigenstate to measure the population of. Can be left
+            undefined if the state's eigenstates form a known eigenbasis with
+            a defined "one state".
+    """
+
+    def __init__(
+        self,
+        *,
+        evaluation_times: Sequence[float] | None = None,
+        one_state: Eigenstate | None = None,
+    ):
+        """Initializes the observable."""
+        super().__init__(evaluation_times=evaluation_times)
+        # TODO: Checks
+        self.one_state = one_state
+
+    def name(self) -> str:
+        """Returns the name of the observable."""
+        return "qubit_density"
+
+    def apply(
+        self, *, state: State, hamiltonian: Operator, **kwargs: Any
+    ) -> list:
+        """Calculates the observable to store in the Results."""
+        return [
+            CorrelationMatrix._get_number_operator(
```

Comment:
Same as with the CorrelationMatrix, except it's less critical here. Maybe we can figure something out